src/layers/BinaryOpLayer.cpp

```diff
@@ -0,0 +1,237 @@
+#include "layers/BinaryOpLayer.hpp"
+
+namespace it_lab_ai {
+
+namespace {
+template <typename T>
+T apply_binary_op(T a, T b, BinaryOpLayer::Operation op) {
+  switch (op) {
+    case BinaryOpLayer::Operation::kMul:
+      return a * b;
+    case BinaryOpLayer::Operation::kAdd:
+      return a + b;
+    case BinaryOpLayer::Operation::kSub:
+      return a - b;
+    default:
+      throw std::runtime_error("Unsupported binary operation");
+  }
+}
+}  // namespace
+
+void BinaryOpLayer::run(const Tensor& input, Tensor& output) {
+  (void)input;
+  (void)output;
+  throw std::runtime_error(
+      "Use run(const Tensor& A, const Tensor& B, Tensor& output) for binary "
+      "operations");
+}
+
+void BinaryOpLayer::run(const Tensor& A, const Tensor& B, Tensor& output) {
+  if (A.get_type() != B.get_type()) {
+    throw std::runtime_error(
+        "BinaryOpLayer: Input tensors must have the same type");
+  }
+
+  if (is_scalar_tensor(B)) {
+    switch (B.get_type()) {
+      case Type::kFloat:
+        run_with_scalar(A, B.as<float>()->at(0), output);
+        return;
+      case Type::kInt:
+        run_with_scalar(A, static_cast<float>(B.as<int>()->at(0)), output);
+        return;
+      default:
+        throw std::runtime_error("Unsupported scalar type");
+    }
+  }
+
+  if (is_scalar_tensor(A)) {
+    switch (A.get_type()) {
+      case Type::kFloat:
+        run_with_scalar(B, A.as<float>()->at(0), output);
+        return;
+      case Type::kInt:
+        run_with_scalar(B, static_cast<float>(A.as<int>()->at(0)), output);
+        return;
+      default:
+        throw std::runtime_error("BinaryOpLayer: Unsupported scalar type");
+    }
+  }
+
+  if (!can_broadcast(A.get_shape(), B.get_shape())) {
+    throw std::runtime_error(
+        "BinaryOpLayer: Incompatible shapes for broadcasting");
+  }
+
+  Shape output_shape =
+      calculate_broadcasted_shape(A.get_shape(), B.get_shape());
+
+  switch (A.get_type()) {
+    case Type::kFloat: {
```

Comment:
I suggest extracting broadcast to `run_broadcast_impl` similar to `run_with_scalar`